test/sslapitest.c

```diff
@@ -5324,13 +5329,22 @@ static int test_shutdown(int tst)
                     * Even though we're shutdown on receive we should still be
                     * able to write.
                     */
-                || !TEST_true(SSL_write(serverssl, msg, sizeof(msg)))
-                || !TEST_int_eq(SSL_shutdown(serverssl), 1))
+                || !TEST_true(SSL_write(serverssl, msg, sizeof(msg))))
+            goto end;
+        if (tst == 4 &&
+                (!TEST_true(SSL_key_update(serverssl, SSL_KEY_UPDATE_NOT_REQUESTED))
```

Comment:
What happens on the client if SSL_KEY_UPDATE_REQUESTED is used? Does it attempt to send data after shutdown has been sent?

Reply:
This doesn't have any effect on the test, it still passes. The documentation says that an SSL_read() can send it, but it doesn't look like it's doing it. I first added a test that also tried that and I expected it to fail somewhere. I haven't figured out yet why it doesn't fail. My current assumption is that it's only send as part of an SSL_write().

Reply:
Or it could actually send the packet, in which case I think there are 2 bugs.

Reply:
It *does* send the packet. Which is what I would have expected after this PR. This should never happen in master before this change: you cannot send a close_notify while in init, and you can never re-enter init if a close_notify was sent (because we ignore incoming handshake messages in that case). Therefore a handshake message will never be sent after a close_notify has been sent. After this change that is no longer true and we get this problem.

Possible solutions:
1) Change things so that if we attempt to write out a handshake message then we just pretend we were successful, but don't actually do it.
2) Change the state machine so that it detects this case and ditch the processing that would lead us to attempt to write something out in the first place.

The problem with (1) is that we could end up in a situation where we never exit init because we attempt to read another handshake message from the peer - but that will never arrive because we didn't actually send it (e.g. in a reneg). (2) is possible but we'll have to make sure we get every place where we could possibly try and write something out otherwise we'll have introduced a bug. I'm quite worried about that as it sounds like a potential route for introducing a whole class of errors.

A third possibility is that we do what we're doing in master now (i.e. ditch the handshake message if we receive it after we've already sent close_notify) *unless* it's one we know we can handle (e.g. NewSessionTicket). If we do that then we should probably move the message ditching code out of the record layer and into the state machine.

It was this problem that made me somewhat reluctant to remove the handshake message ditching code in the first place. It's a whole lot easier if you can just make the simplifying assumption that we just ditch all the handshake messages. If we don't think that is reasonable, then (3) is the way to go IMO.